### PR TITLE
Add much of the remaining documentation for the existing Hail2 API.

### PR DESCRIPTION
**Pull request by tpoterba**

### CODE REVIEW

**Comment by tpoterba** [comment]
Let me know if you don't have bandwidth to review and I'll assign someone else.

File: python/hail/api2/table.py [hunk 1]
```diff
@@ -444,33 +538,85 @@ def filter(self, expr, keep=True):
...
-        >>> table_result = table1.select(table1.C1)
+        Note
+        ----
 
-        Reorder the columns:
+        **Using select**
 
-        >>> table_result = table1.select(table1.C3, table1.C1, table1.C2)
+        Select and its sibling methods (:meth:`Table.select_globals`,
+        :meth:`MatrixTable.select_globals`, :meth:`MatrixTable.select_rows`,
```
Comment (konradjk) [inline python/hail/api2/table.py]: Mixing `Table` and `MatrixTable` here intentionally?

File: python/hail/api2/table.py [hunk 2]
```diff
@@ -529,23 +708,129 @@ def drop(self, *exprs):
...
+        Examples
+        --------
+        Export to a tab-separated file:
+
+        >>> table1.export('output/table1.tsv.bgz')
+
+        Note
+        ----
+        It is highly recommended to export large files with a ``.bgz`` extension,
+        which will use a block gzipped compression codec.
```
Comment (konradjk) [inline python/hail/api2/table.py]: Might suggest: "which can be read natively with Python's `gzip.open` and R's `read.table` natively"

File: python/hail/api2/table.py [hunk 3]
```diff
@@ -529,23 +708,129 @@ def drop(self, *exprs):
...
+
+        Parameters
+        ----------
+        output : str
+            URI at which to write exported file.
+        types_file : str
```
Comment (konradjk) [inline python/hail/api2/table.py]: or None

File: python/hail/api2/matrixtable.py [hunk 4]
```diff
@@ -263,55 +340,181 @@ def _jvdf(self):
...
+        Add two global fields:
+
+        >>> pops_1kg = {'EUR', 'AFR', 'EAS', 'SAS', 'AMR'}
+        >>> dataset_result = dataset.annotate_globals(pops_in_1kg = pops_1kg,
+        ...                                           gene_list = ['SHH', 'SCN1A', 'SPTA1', 'DISC1'])
+
+        Add global fields from another table and matrix table:
+
+        >>> dataset_result = dataset.annotate_globals(thing1 = dataset2[:, :].global_field,
+        ...                                           thing2 = table1[:].global_field)
```
Comment (cseed) [inline python/hail/api2/matrixtable.py]: Are the slices here necessary?  Should `annotate_globals` be able to take a indexless expression even if it is from another object?

File: python/hail/api2/matrixtable.py [hunk 5]
```diff
@@ -5,6 +5,11 @@
 from hail.api2 import Table
 
 class GroupedMatrixTable(object):
+    """Matrix table grouped by row or column that can be aggregated to produce a new matrix table.
+
+    There are only two operations on a grouped matrix table, :meth:`GroupedMatrixTable.set_partitions`
```
[the remark this answers is not in the document]
Comment (tpoterba) [reply python/hail/api2/matrixtable.py]: I think at some point we'll do our own optimizations, and may decide your hint is wrong (like if you do `table.group_by().hint_partitions(1000).aggregate(...)`, then I know there's only one row in the resulting table, and should set the result partitions to 1.

File: python/hail/api2/matrixtable.py [hunk 6]
```diff
@@ -15,7 +20,45 @@ def __init__(self, parent, group, grouped_indices):
...
+        --------
+
+        Use `partition_hint` in a :meth:`MatrixTable.group_rows_by` /
+        :meth:`GroupedMatrixTable.aggregate` pipeline:
+
+        .. testsetup::
+
+            dataset = dataset.annotate_rows(gene = 'gene')
+
+        >>> dataset_result = dataset.group_rows_by(dataset.gene)
```
[the remark this answers is not in the document]
Comment (tpoterba) [reply python/hail/api2/matrixtable.py]: yup, good call. The backslashes caused problems rendering without the doctests, but no parens fails tests.

File: python/hail/api2/matrixtable.py [hunk 7]
```diff
@@ -65,6 +132,16 @@ class MatrixTable(object):
         import hail2 as h2
         from hail2 import *
         hc = h2.HailContext()
+        dataset = (vds.annotate_samples_expr('sa = merge(drop(sa, qc), {sample_qc: sa.qc})')
+                      .annotate_variants_expr('va = merge(drop(va, qc), {variant_qc: va.qc})').to_hail2())
+
+        dataset2 = dataset.annotate_globals(global_field=5)
+        table1 = dataset.rows_table()
+        table1 = table1.annotate_globals(global_field=5)
```
[the remark this answers is not in the document]
Comment (tpoterba) [reply python/hail/api2/matrixtable.py]: I use these for doing joins, I think

File: python/hail/api2/matrixtable.py [hunk 8]
```diff
@@ -263,55 +340,181 @@ def _jvdf(self):
...
+        :class:`hail.expr.TStruct`
+            Global schema.
+        """
         if self._global_schema is None:
             self._global_schema = Type._from_java(self._jvds.globalSignature())
         return self._global_schema
 
     @property
     def colkey_schema(self):
+        """The schema of the column key.
```
[the remark this answers is not in the document]
Comment (tpoterba) [reply python/hail/api2/matrixtable.py]: The name is always `s` for now. I want to flesh out the docs when we fix the infrastructure to open that up.

File: python/hail/api2/matrixtable.py [hunk 9]
```diff
@@ -263,55 +340,181 @@ def _jvdf(self):
...
     @handle_py4j
     def annotate_rows(self, **named_exprs):
+        """Create new row-indexed fields by name.
+
+        Examples
+        --------
+        Compute call statistics for high quality samples per variant:
+
+        >>> high_quality_calls = f.filter(dataset.GT, dataset.sample_qc.meanGQ > 20)
+        >>> dataset_result = dataset.annotate_rows(call_stats = f.call_stats(high_quality_calls, dataset.v)
```
[the remark this answers is not in the document]
Comment (tpoterba) [reply python/hail/api2/matrixtable.py]: wow, good catch. Sphinx caught that for me earlier.

File: python/hail/api2/matrixtable.py [hunk 10]
```diff
@@ -263,55 +340,181 @@ def _jvdf(self):
...
+        :class:`hail.expr.Type`
+             Column key schema.
+        """
         if self._colkey_schema is None:
             self._colkey_schema = Type._from_java(self._jvds.sSignature())
         return self._colkey_schema
 
     @property
     def col_schema(self):
+        """The schema of column-indexed fields in the matrix.
```
[the remark this answers is not in the document]
Comment (tpoterba) [reply python/hail/api2/matrixtable.py]: I think we should write a doc separately to detail how 0.2 maps onto 0.1. Made a note of that

File: python/hail/api2/matrixtable.py [hunk 11]
```diff
@@ -263,55 +340,181 @@ def _jvdf(self):
...
+        :class:`MatrixTable`.
+
+        >>> dataset_result = dataset.annotate_rows(consequence = table1[dataset.v].consequence,
+        ...                                        dataset2_AF = dataset2[dataset.v, :].info.AF)
+
+        Note
+        ----
+        This method supports aggregation over columns. For instance, the usage:
+
+        >>> dataset_result = dataset.select_rows(mean_GQ = f.mean(dataset.GQ))
```
[the remark this answers is not in the document]
Comment (tpoterba) [reply python/hail/api2/matrixtable.py]: yep, definitely did.

File: python/hail/api2/matrixtable.py [hunk 12]
```diff
@@ -375,7 +702,48 @@ def select_globals(self, *exprs, **named_exprs):
...
+        with a differently-indexed field of the table, the method will fail.
+
+        Note
+        ----
+
+        See :py:meth:`Table.select` for more information about using ``select`` methods.
+
+        Note
+        ----
+        This method supports aggregation over rows. For instance, the usage:
```
[the remark this answers is not in the document]
Comment (tpoterba) [reply python/hail/api2/matrixtable.py]: can you explain a bit more? We definitely need to clearly describe the differences, but my thought is that:

**annotate** adds fields to the schema
**select** adds the fields given into an empty schema
**transmute** is like **annotate** followed by **drop**ping every field referenced in an expression

File: python/hail/api2/matrixtable.py [hunk 13]
```diff
@@ -401,7 +769,46 @@ def select_cols(self, *exprs, **named_exprs):
...
     def select_rows(self, *exprs, **named_exprs):
-        exprs = tuple(to_expr(e) for e in exprs)
+        """Select existing row fields or create new fields by name, dropping the rest.
+
+        Examples
+        --------
+        Select existing fields and compute a new one:
+
+        >>> dataset_result = dataset.select_rows(dataset.variant_qc.meanGQ,
+        ...                                      highQualityCases = f.count_where((dataset.GQ > 20) & (dataset.isCase)))
```
[the remark this answers is not in the document]
Comment (tpoterba) [reply python/hail/api2/matrixtable.py]: nope, can't override the `and` operator. Unfortunately all these parens are necessary :(

[the remark this answers is not in the document]
Comment (tpoterba) [reply python/hail/api2/matrixtable.py]: & binds more highly than `and`. It is very annoying.

File: python/hail/api2/matrixtable.py [hunk 14]
```diff
@@ -450,7 +891,44 @@ def select_entries(self, *exprs, **named_exprs):
     @handle_py4j
     @typecheck_method(exprs=tupleof(oneof(strlike, Expression)))
     def drop(self, *exprs):
-        """Drop fields from the matrix."""
+        """Drop fields.
```
[the remark this answers is not in the document]
Comment (tpoterba) [reply python/hail/api2/matrixtable.py]: yep, good call

File: python/hail/api2/matrixtable.py [hunk 15]
```diff
@@ -493,46 +971,298 @@ def drop(self, *exprs):
...
+        ``True``). If `keep` is ``False``, then entries where `expr` evaluates to
+        ``False`` will be removed (the filter removes the entries where the predicate
+        evaluates to ``True``.
+
+        Note
+        ----
+        "Removal" of an entry constitutes setting all its fields to missing. There
+        is some debate about what removing an entry of a matrix means semantically,
+        given the representation of a :class:`MatrixTable` as a whole workspace in
+        Hail.
```
[the remark this answers is not in the document]
Comment (tpoterba) [reply python/hail/api2/matrixtable.py]: yeah, I'm not sure. We started discussing this on slack yesterday, we'll loop you in when we decide to dig deeper.

Entries are sorta special.

File: python/hail/api2/matrixtable.py [hunk 16]
```diff
@@ -493,46 +971,298 @@ def drop(self, *exprs):
...
+
+        Examples
+        --------
+        Aggregate over rows:
+
+        .. doctest::
+
+            >>> dataset.aggregate_rows(n_high_quality=f.count_where(dataset.qual > 40),
+            ...                        mean_qual = f.mean(dataset.qual))
+            Struct(n_high_quality=0.6109954, mean_qual=50.12515572)
```
[the remark this answers is not in the document]
Comment (tpoterba) [reply python/hail/api2/matrixtable.py]: good catch. I just made up realistic ish outputs and must have though I used fraction

File: python/hail/api2/matrixtable.py [hunk 17]
```diff
@@ -493,46 +971,298 @@ def drop(self, *exprs):
...
+        This method supports (and expects!) aggregation over rows.
+
+        Parameters
+        ----------
+        named_exprs : keyword args of :class:`hail.expr.expression.Expression`
+            Aggregation expressions.
+
+        Returns
+        -------
+        :class:`Struct`
```
[the remark this answers is not in the document]
Comment (tpoterba) [reply python/hail/api2/matrixtable.py]: nope, I agree this is somewhat confusing.

Table.aggregate, MT.aggregate_{rows, cols, entries} return structs.

Table.group_by().aggregate returns a table
MT.group_{rows, cols}_by().aggregate returns a MT

[the remark this answers is not in the document]
Comment (tpoterba) [reply python/hail/api2/matrixtable.py]: the alternative returns a struct as well

File: python/hail/api2/matrixtable.py [hunk 18]
```diff
@@ -592,77 +1397,312 @@ def aggregate_entries(self, **named_exprs):
...
-            elif self._fields[expr].indices != self._row_indices:
+    def explode_rows(self, field_expr):
+        """Explodes a row field of type array or set, copying the entire row for each element.
+
+        Examples
+        --------
+
+        .. testsetup::
+
+            dataset = dataset.annotate_rows(gene=['TITIN'])
```
[the remark this answers is not in the document]
Comment (tpoterba) [reply python/hail/api2/matrixtable.py]: will fix, even though this is never actually rendered in the docs 😛

File: python/hail/api2/matrixtable.py [hunk 19]
```diff
@@ -592,77 +1397,312 @@ def aggregate_entries(self, **named_exprs):
...
+
+        >>> dataset_result = dataset.explode_rows(dataset.gene)
+
+        Notes
+        -----
+        The new matrix table will have `N` copies of each row, where `N` is the number
+        of elements that row contains for the field denoted by `field_expr`. The field
+        referenced in `field_expr` is replaced in the sequence of duplicated rows by the
+        sequence of elements in the array or set. All other fields remain the same,
+        including entry fields.
```
[the remark this answers is not in the document]
Comment (tpoterba) [reply python/hail/api2/matrixtable.py]: good idea. I'm not going to do this now, but we will do in another PR soon.

File: python/hail/api2/table.py (hunk 1 above)
Comment (tpoterba) [reply python/hail/api2/table.py]: yep! Want to show that all these methods are similar.

File: python/hail/api2/matrixtable.py (hunk 4 above)
Comment (tpoterba) [reply python/hail/api2/matrixtable.py]: this is a longer conversation. I think yes, we do need the `[:]`. This permits us to have drastically simplified expression unification logic, and probably nicer error messages when the user is doing something wrong.

File: python/hail/api2/matrixtable.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (konradjk) [reply python/hail/api2/matrixtable.py]: I see. Ok, that makes sense. And they can all aggregate?

[the remark this answers is not in the document]
Comment (tpoterba) [reply python/hail/api2/matrixtable.py]: yep. The expressions permitted are the same, except for select (which also has varargs for existing fields)

File: python/hail/api2/matrixtable.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (konradjk) [reply python/hail/api2/matrixtable.py]: Yeah, sorry, I missed that these were in the .. testsetup:: not in the examples

**Pull request merged**